File: Python/Euler/PE006.py
"""
| **Problem 6:** Sum square difference
| URL: https://projecteuler.net/problem=6
::

    The sum of the squares of the first ten natural numbers is,
    1^2 + 2^2 + ... + 10^2 = 385

    The square of the sum of the first ten natural numbers is,
    (1 + 2 + ... + 10)^2 = 552 = 3025

    Hence, the difference between the sum of the squares of
    the first ten natural numbers and the square of the sum
    is 3025 - 385 = 2640.

    Find the difference between the sum of the squares of the
    first one hundred natural numbers and the square of the sum.

:author: Gorka Suárez García
:copyright: (c) 2022, Gorka Suárez García
"""

######################################################################
# Constants
######################################################################

FIRST = 1
LAST = 100


######################################################################
# Functions
######################################################################

def calc_number(first, last):
    """
    Calcs the difference between the sum of the squares
    and the square of the sum of a range of numbers.
    :param first: The first value of the range.
    :param last: The last value of the range.
    :return: The difference between the sums.
    """
    numbers = range(first, last + 1)
    squares = [number ** 2 for number in numbers]
    sum_square = sum(squares)
    square_sum = sum(numbers) ** 2
    return square_sum - sum_square


def main():
    """
    Main entry for the problem solver.
    """
    result = calc_number(FIRST, LAST)

    # Show the final result of the problem:
    message = "The difference between the sum of the squares "
    message += "and the square of the sum of the first "
    message += f"{LAST} numbers is {result}."
    print(message)


######################################################################
# Module execution
######################################################################

if __name__ == "__main__":
    main()

File: CSharp/Euler/PE006.cs
//======================================================================
// Copyright (C) 2023, Gorka Suárez García
//======================================================================

/*
 * The sum of the squares of the first ten natural numbers is,
 * 1^2 + 2^2 + ... + 10^2 = 385
 * 
 * The square of the sum of the first ten natural numbers is,
 * (1 + 2 + ... + 10)^2 = 552 = 3025
 * 
 * Hence, the difference between the sum of the squares of
 * the first ten natural numbers and the square of the sum
 * is 3025 - 385 = 2640.
 * 
 * Find the difference between the sum of the squares of the
 * first one hundred natural numbers and the square of the sum.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Euler {
    /// <summary>
    /// This class represents the problem 6.
    /// </summary>
    public class PE006 : IRunnable {
        /// <summary>
        /// Main entry for the problem solver.
        /// </summary>
        public void Run () {
            const int FIRST = 1, LAST = 100;

            var result = CalcNumber(FIRST, LAST);

            Console.Write("The difference between the sum of the squares and the square ");
            Console.WriteLine($"of the sum of the first {LAST} numbers is {result}.");
        }

        /// <summary>
        /// Calcs the difference between the sum of the squares
        /// and the square of the sum of a range of numbers.
        /// </summary>
        /// <param name="first">The first value of the range.</param>
        /// <param name="last">The last value of the range.</param>
        /// <returns>The difference between the sums.</returns>
        public long CalcNumber (int first, int last) {
            Func<long, long> square = x => x * x;
            var numbers = Tools.Range<long>(first, last);
            var sum_square = numbers.Select(square).Sum();
            var square_sum = square(numbers.Sum());
            return square_sum - sum_square;
        }
    }
}
